File: packages/game/engine/src/runtime/gameObjectStore.ts
import {
  LocationId,
  WorldId,
} from '@pixsim7/shared.types';
import type {
  GameObject,
  GameObjectCapabilityId,
  GameObjectId,
  GameObjectStore,
  GameSessionDTO,
  Transform,
} from '@pixsim7/shared.types';
import { GameObjectEntity } from './GameObjectEntity';

export const GAME_OBJECT_STORE_SCHEMA_VERSION = 1;

export interface GameObjectQuery {
  kind?: string;
  locationId?: number;
  capability?: GameObjectCapabilityId | string;
  tags?: string[];
}

export type GameObjectLookup =
  | string
  | {
      kind: string;
      id: GameObjectId;
    };

function asRecord(value: unknown): Record<string, unknown> | null {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return null;
  }
  return value as Record<string, unknown>;
}

function toNumber(value: unknown): number | null {
  const numeric = typeof value === 'number' ? value : Number(value);
  return Number.isFinite(numeric) ? numeric : null;
}

function createFallbackTransform(session: Pick<GameSessionDTO, 'world_id'>, locationId?: number): Transform {
  const worldId = toNumber(session.world_id) ?? 0;
  const normalizedLocationId = toNumber(locationId);
  const transform: Transform = {
    worldId: WorldId(worldId),
    position: { x: 0, y: 0 },
    space: 'world_2d',
  };
  if (normalizedLocationId !== null && normalizedLocationId >= 0) {
    transform.locationId = LocationId(normalizedLocationId);
  }
  return transform;
}

function normalizeTransform(
  rawTransform: unknown,
  fallback: Transform
): Transform {
  const record = asRecord(rawTransform);
  const position = asRecord(record?.position);
  const posX = toNumber(position?.x);
  const posY = toNumber(position?.y);
  const worldId = toNumber(record?.worldId);
  if (record && position && posX !== null && posY !== null && worldId !== null) {
    const normalized: Transform = {
      ...fallback,
      ...(record as unknown as Transform),
      worldId: WorldId(worldId),
      position: {
        ...(position as unknown as Transform['position']),
        x: posX,
        y: posY,
      },
    };
    const locationId = toNumber(record.locationId);
    if (locationId !== null && locationId >= 0) {
      normalized.locationId = LocationId(locationId);
    }
    return normalized;
  }
  return fallback;
}

function normalizeCapabilityId(value: unknown): string | null {
  if (typeof value !== 'string') return null;
  const trimmed = value.trim();
  return trimmed.length > 0 ? trimmed : null;
}

export function toGameObjectRef(kind: string, id: GameObjectId): string {
  const normalizedKind = typeof kind === 'string' ? kind.trim() : '';
  const normalizedId =
    typeof id === 'string'
      ? id.trim()
      : typeof id === 'number' && Number.isFinite(id)
      ? String(id)
      : '';
  if (!normalizedKind || !normalizedId) {
    throw new Error('Cannot build game object ref without kind and id');
  }
  return `${normalizedKind}:${normalizedId}`;
}

function normalizeGameObject(
  raw: unknown,
  fallbackRef: string,
  session: Pick<GameSessionDTO, 'world_id'>
): GameObject | null {
  const record = asRecord(raw);
  if (!record) return null;
  const kind = typeof record.kind === 'string' ? record.kind.trim() : null;
  const idValue = record.id;
  if (!kind || (typeof idValue !== 'string' && typeof idValue !== 'number')) {
    return null;
  }
  const normalizedStringId =
    typeof idValue === 'string' ? idValue.trim() : null;
  const normalizedNumericId =
    typeof idValue === 'number' && Number.isFinite(idValue) ? idValue : null;
  if (normalizedStringId !== null && normalizedStringId.length === 0) {
    return null;
  }
  if (normalizedStringId === null && normalizedNumericId === null) {
    return null;
  }
  const id = (normalizedStringId ?? normalizedNumericId) as GameObjectId;
  const ref =
    typeof record.ref === 'string' && record.ref.trim().length > 0
      ? record.ref
      : toGameObjectRef(kind, id);
  const fallback = createFallbackTransform(session);
  const transform = normalizeTransform(record.transform, fallback);
  const capabilities = Array.isArray(record.capabilities)
    ? record.capabilities
        .map((entry) => {
          const capabilityRecord = asRecord(entry);
          const capabilityId = normalizeCapabilityId(capabilityRecord?.id);
          if (!capabilityId) return null;
          return {
            ...(capabilityRecord as Record<string, unknown>),
            id: capabilityId,
            enabled: capabilityRecord?.enabled !== false,
          };
        })
        .filter((entry): entry is NonNullable<typeof entry> => entry !== null)
    : undefined;
  const normalizedRecord: Record<string, unknown> = {
    ...record,
    kind,
    id,
    ref,
    runtimeKind:
      typeof record.runtimeKind === 'string' && record.runtimeKind.trim().length > 0
        ? record.runtimeKind
        : kind,
    name:
      typeof record.name === 'string' && record.name.trim().length > 0
        ? record.name
        : fallbackRef,
    transform,
    capabilities,
  };
  return normalizedRecord as unknown as GameObject;
}

/**
 * Build a canonical item-kind `GameObject` for an inventory item.
 */
export function buildInventoryItemObject(
  session: Pick<GameSessionDTO, 'world_id'>,
  itemId: string,
  quantity: number,
  metadata?: Record<string, unknown>
): GameObject {
  const id = itemId.trim();
  const qty = Number.isFinite(Number(quantity)) ? Math.max(0, Number(quantity)) : 0;
  const meta = asRecord(metadata) ?? {};
  const name =
    typeof meta.name === 'string' && meta.name.trim().length > 0 ? meta.name : id;
  const extraItemData = asRecord(meta.itemData) ?? {};
  const reservedMetaKeys = new Set(['name', 'itemData', 'id', 'itemId', 'qty', 'quantity']);
  const restMeta: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(meta)) {
    if (!reservedMetaKeys.has(key)) restMeta[key] = value;
  }
  return {
    kind: 'item',
    id,
    ref: `item:${id}`,
    name,
    runtimeKind: 'item',
    transform: createFallbackTransform(session),
    capabilities: [{ id: 'inventory_item', enabled: true }],
    itemData: {
      ...restMeta,
      ...extraItemData,
      itemDefId: id,
      quantity: qty,
    },
    meta: {
      source: 'canonical.inventory',
    },
  };
}

/**
 * Build a canonical `npc`-kind `GameObject` shell (no components yet).
 *
 * Mirrors `buildInventoryItemObject` for the npc domain: narrative / ECS
 * component state lives on the object's `components[]` array, keyed by
 * component `type` (e.g. `narrative`), matching the backend's
 * `npc.components[type='stats:<def_id>']` convention.
 */
export function buildNpcObject(
  session: Pick<GameSessionDTO, 'world_id'>,
  npcId: number | string,
  metadata?: Record<string, unknown>
): GameObject {
  const id = String(npcId).trim();
  const meta = asRecord(metadata) ?? {};
  const name =
    typeof meta.name === 'string' && meta.name.trim().length > 0 ? meta.name : `npc:${id}`;
  return {
    kind: 'npc',
    id,
    ref: `npc:${id}`,
    name,
    runtimeKind: 'npc',
    transform: createFallbackTransform(session),
    capabilities: [{ id: 'narrative_participant', enabled: true }],
    components: [],
    meta: {
      source: 'canonical.npc',
    },
  };
}

function normalizeStore(
  rawStore: unknown,
  session: Pick<GameSessionDTO, 'world_id'>
): GameObjectStore | null {
  const record = asRecord(rawStore);
  const objectsRecord = asRecord(record?.objects);
  if (!record || !objectsRecord) return null;

  const objects: Record<string, GameObject> = {};
  for (const [ref, rawObject] of Object.entries(objectsRecord)) {
    const normalized = normalizeGameObject(rawObject, ref, session);
    if (!normalized) continue;
    const normalizedRef = typeof normalized.ref === 'string' ? normalized.ref : ref;
    objects[normalizedRef] = normalized;
  }

  return {
    schemaVersion: toNumber(record.schemaVersion) ?? GAME_OBJECT_STORE_SCHEMA_VERSION,
    objects,
    meta: asRecord(record.meta) ?? undefined,
  };
}

export function getSessionGameObjectStore(
  session: Pick<GameSessionDTO, 'world_id' | 'flags'>
): GameObjectStore {
  const canonical = normalizeStore(asRecord(session.flags)?.gameObjects, session);
  return (
    canonical ?? {
      schemaVersion: GAME_OBJECT_STORE_SCHEMA_VERSION,
      objects: {},
    }
  );
}

function hasCapability(object: GameObject, capability: string): boolean {
  if (!capability) return true;
  const required = capability.trim();
  if (!required) return true;
  const capabilities = object.capabilities ?? [];
  return capabilities.some((item) => item.id === required && item.enabled !== false);
}

function hasAllTags(object: GameObject, tags: string[]): boolean {
  if (tags.length === 0) return true;
  const objectTags = new Set((object.tags ?? []).map((tag) => tag.toLowerCase()));
  return tags.every((tag) => objectTags.has(tag.toLowerCase()));
}

function matchesQuery(object: GameObject, query: GameObjectQuery): boolean {
  if (query.kind && object.kind !== query.kind) {
    return false;
  }
  if (query.locationId != null) {
    const locationId = toNumber(object.transform?.locationId);
    if (locationId !== query.locationId) {
      return false;
    }
  }
  if (query.capability && !hasCapability(object, query.capability)) {
    return false;
  }
  if (query.tags && !hasAllTags(object, query.tags)) {
    return false;
  }
  return true;
}

export function listSessionGameObjects(
  session: Pick<GameSessionDTO, 'world_id' | 'flags'>,
  query: GameObjectQuery = {}
): GameObject[] {
  const store = getSessionGameObjectStore(session);
  return Object.entries(store.objects)
    .sort(([left], [right]) => left.localeCompare(right))
    .map(([, object]) => object)
    .filter((object) => matchesQuery(object, query));
}

export function getSessionGameObject(
  session: Pick<GameSessionDTO, 'world_id' | 'flags'>,
  lookup: GameObjectLookup
): GameObject | null {
  const store = getSessionGameObjectStore(session);
  const ref =
    typeof lookup === 'string'
      ? lookup
      : toGameObjectRef(lookup.kind, lookup.id);
  return store.objects[ref] ?? null;
}

/**
 * Hydration -> object-core seam (migration_rollout_v1 / hydration-entry).
 *
 * The single place canonical session state (`flags.gameObjects`) is projected
 * into runtime `GameObjectEntity` instances. Runtime callers should consume
 * entities via these (or the
 * `GameRuntime` accessors that delegate here); the plain-POJO functions stay
 * the persistence/API edge per the POJO-boundary policy.
 */
export function listSessionGameObjectEntities(
  session: Pick<GameSessionDTO, 'world_id' | 'flags'>,
  query: GameObjectQuery = {}
): GameObjectEntity[] {
  return GameObjectEntity.fromPOJOs(listSessionGameObjects(session, query));
}

export function getSessionGameObjectEntity(
  session: Pick<GameSessionDTO, 'world_id' | 'flags'>,
  lookup: GameObjectLookup
): GameObjectEntity | null {
  const pojo = getSessionGameObject(session, lookup);
  return pojo ? GameObjectEntity.fromPOJO(pojo) : null;
}

/**
 * Write a merged object set back into the session as the canonical
 * `flags.gameObjects` store. As of the canonical cutover, the legacy
 * `flags.inventory.items` mirror is no longer maintained — both sides read
 * canonical item GameObjects directly.
 */
function writeStoreObjects(
  session: GameSessionDTO,
  mergedObjects: Record<string, GameObject>,
  baseStore: GameObjectStore
): GameSessionDTO {
  const currentFlags = asRecord(session.flags) ?? {};
  const nextStore: GameObjectStore = {
    schemaVersion: GAME_OBJECT_STORE_SCHEMA_VERSION,
    objects: mergedObjects,
    meta: {
      ...(asRecord(baseStore.meta) ?? {}),
      updatedAt: new Date().toISOString(),
    },
  };

  return {
    ...session,
    flags: {
      ...currentFlags,
      gameObjects: nextStore,
    },
  };
}

export function upsertSessionGameObjects(
  session: GameSessionDTO,
  objects: GameObject[]
): GameSessionDTO {
  if (!Array.isArray(objects) || objects.length === 0) {
    return session;
  }

  const baseStore = getSessionGameObjectStore(session);
  const mergedObjects: Record<string, GameObject> = { ...baseStore.objects };
  for (const object of objects) {
    const fallbackRef = toGameObjectRef(object.kind, object.id);
    const normalized = normalizeGameObject(object, fallbackRef, session);
    if (!normalized) continue;
    const key = typeof normalized.ref === 'string' ? normalized.ref : fallbackRef;
    mergedObjects[key] = normalized;
  }

  return writeStoreObjects(session, mergedObjects, baseStore);
}

/**
 * Remove canonical game objects by ref from the `flags.gameObjects` store.
 */
export function removeSessionGameObjects(
  session: GameSessionDTO,
  refs: string[]
): GameSessionDTO {
  if (!Array.isArray(refs) || refs.length === 0) {
    return session;
  }

  const baseStore = getSessionGameObjectStore(session);
  const mergedObjects: Record<string, GameObject> = { ...baseStore.objects };
  let changed = false;
  for (const ref of refs) {
    if (ref in mergedObjects) {
      delete mergedObjects[ref];
      changed = true;
    }
  }

  if (!changed) {
    return session;
  }

  return writeStoreObjects(session, mergedObjects, baseStore);
}

// =============================================================================
// NPC component accessors (canonical npc-kind GameObject `components[]`)
// =============================================================================
//
// NPC component state (narrative runtime state, ad-hoc ECS components from
// narrative effects) lives on the canonical npc GameObject's `components[]`
// array, keyed by component `type`. These helpers are the single read/write
// seam; legacy `flags.npcs[*].components` is no longer written.

/** Read an npc component's `data` payload (canonical). Returns null if absent. */
export function getNpcComponentData(
  session: Pick<GameSessionDTO, 'world_id' | 'flags'>,
  npcId: number | string,
  type: string
): Record<string, unknown> | null {
  const object = getSessionGameObject(session, `npc:${String(npcId)}`);
  if (!object) return null;
  const component = (object.components ?? []).find(
    (comp) => comp.type === type && comp.enabled !== false
  );
  return asRecord(component?.data);
}

/**
 * Upsert an npc component's `data` (immutable). Creates the npc GameObject if it
 * does not exist, then replaces (by `type`) or appends the component entry.
 */
export function upsertNpcComponent(
  session: GameSessionDTO,
  npcId: number | string,
  type: string,
  data: Record<string, unknown>
): GameSessionDTO {
  const ref = `npc:${String(npcId)}`;
  const existing = getSessionGameObject(session, ref);
  const base = existing ?? buildNpcObject(session, npcId);
  const components = [...(base.components ?? [])];
  const index = components.findIndex((comp) => comp.type === type);
  const nextComponent = { type, enabled: true, data };
  if (index >= 0) {
    components[index] = { ...components[index], ...nextComponent };
  } else {
    components.push(nextComponent);
  }
  return upsertSessionGameObjects(session, [{ ...base, components }]);
}

/**
 * Remove an npc component by `type` (immutable). The npc GameObject itself is
 * left in place even if it becomes component-less.
 */
export function removeNpcComponent(
  session: GameSessionDTO,
  npcId: number | string,
  type: string
): GameSessionDTO {
  const ref = `npc:${String(npcId)}`;
  const existing = getSessionGameObject(session, ref);
  if (!existing) return session;
  const components = (existing.components ?? []).filter((comp) => comp.type !== type);
  if (components.length === (existing.components ?? []).length) {
    return session;
  }
  return upsertSessionGameObjects(session, [{ ...existing, components }]);
}

File: pixsim7/backend/main/services/game/game_object_store.py
"""
Backend runtime GameObject store.

Python port of the TypeScript engine store
(`packages/game/engine/src/runtime/gameObjectStore.ts`). Operates over the
canonical `session.flags["gameObjects"]` shape defined by the shared contract
(`packages/shared/types/src/game.ts`): a `GameObjectStore` of
`{schemaVersion, objects: {<ref>: GameObject}, meta}` keyed by canonical ref
(e.g. "item:flower", "npc:12").

Keys are camelCase to match the frontend-authored JSON shape verbatim
(`gameObjects`, `itemData`, `runtimeKind`, `worldId`, ...).

Scope (backend-canonical-gameobject-adoption, checkpoint 1): the store itself
plus read/list/get/upsert/remove and the TEMPORARY `flags.inventory` mirror.
It is greenfield and not yet wired into InventoryService / narrative ECS — that
is checkpoints 2 and 3. Behaviour intentionally mirrors the TS store so the two
runtimes stay parity-testable.
"""
from __future__ import annotations

import math
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, Union

GAME_OBJECT_STORE_SCHEMA_VERSION = 1

GameObjectLookup = Union[str, Dict[str, Any]]


# ---------------------------------------------------------------------------
# Low-level coercion helpers (mirror the TS asRecord/toNumber/toStringArray)
# ---------------------------------------------------------------------------

def _as_record(value: Any) -> Optional[Dict[str, Any]]:
    return value if isinstance(value, dict) else None


def _is_finite_number(value: Any) -> bool:
    if isinstance(value, bool):
        return False
    if isinstance(value, int):
        return True
    if isinstance(value, float):
        return math.isfinite(value)
    return False


def _to_number(value: Any) -> Optional[float]:
    if isinstance(value, bool):
        return None
    if isinstance(value, (int, float)):
        return float(value) if math.isfinite(float(value)) else None
    if isinstance(value, str):
        try:
            parsed = float(value)
        except (TypeError, ValueError):
            return None
        return parsed if math.isfinite(parsed) else None
    return None


def _id_to_ref_string(obj_id: Any) -> str:
    if isinstance(obj_id, str):
        return obj_id.strip()
    if isinstance(obj_id, bool):
        return ""
    if isinstance(obj_id, (int, float)) and math.isfinite(float(obj_id)):
        as_float = float(obj_id)
        return str(int(as_float)) if as_float.is_integer() else str(obj_id)
    return ""


def to_game_object_ref(kind: Any, obj_id: Any) -> str:
    normalized_kind = kind.strip() if isinstance(kind, str) else ""
    normalized_id = _id_to_ref_string(obj_id)
    if not normalized_kind or not normalized_id:
        raise ValueError("Cannot build game object ref without kind and id")
    return f"{normalized_kind}:{normalized_id}"


# ---------------------------------------------------------------------------
# Transform / object normalization
# ---------------------------------------------------------------------------

def _create_fallback_transform(world_id: Any, location_id: Any = None) -> Dict[str, Any]:
    world = _to_number(world_id)
    transform: Dict[str, Any] = {
        "worldId": int(world) if world is not None else 0,
        "position": {"x": 0, "y": 0},
        "space": "world_2d",
    }
    location = _to_number(location_id)
    if location is not None and location >= 0:
        transform["locationId"] = int(location)
    return transform


def _normalize_transform(raw: Any, fallback: Dict[str, Any]) -> Dict[str, Any]:
    record = _as_record(raw)
    position = _as_record(record.get("position")) if record else None
    pos_x = _to_number(position.get("x")) if position else None
    pos_y = _to_number(position.get("y")) if position else None
    world_id = _to_number(record.get("worldId")) if record else None
    if record and position and pos_x is not None and pos_y is not None and world_id is not None:
        normalized: Dict[str, Any] = {**fallback, **record, "worldId": int(world_id)}
        normalized["position"] = {**position, "x": pos_x, "y": pos_y}
        location = _to_number(record.get("locationId"))
        if location is not None and location >= 0:
            normalized["locationId"] = int(location)
        return normalized
    return fallback


def _normalize_capabilities(raw: Any) -> Optional[List[Dict[str, Any]]]:
    if not isinstance(raw, list):
        return None
    result: List[Dict[str, Any]] = []
    for entry in raw:
        record = _as_record(entry)
        cap_id = record.get("id") if record else None
        cap_id = cap_id.strip() if isinstance(cap_id, str) and cap_id.strip() else None
        if not cap_id:
            continue
        result.append({**record, "id": cap_id, "enabled": record.get("enabled") is not False})
    return result


def _normalize_game_object(
    raw: Any, fallback_ref: str, world_id: Any
) -> Optional[Dict[str, Any]]:
    record = _as_record(raw)
    if not record:
        return None
    kind = record.get("kind")
    kind = kind.strip() if isinstance(kind, str) else None
    id_value = record.get("id")
    if not kind:
        return None
    if isinstance(id_value, str):
        normalized_id: Any = id_value.strip()
        if not normalized_id:
            return None
    elif isinstance(id_value, (int, float)) and not isinstance(id_value, bool) and math.isfinite(float(id_value)):
        normalized_id = id_value
    else:
        return None

    ref_raw = record.get("ref")
    ref = ref_raw if isinstance(ref_raw, str) and ref_raw.strip() else to_game_object_ref(kind, normalized_id)
    fallback = _create_fallback_transform(world_id)
    transform = _normalize_transform(record.get("transform"), fallback)
    capabilities = _normalize_capabilities(record.get("capabilities"))

    runtime_kind = record.get("runtimeKind")
    runtime_kind = runtime_kind if isinstance(runtime_kind, str) and runtime_kind.strip() else kind
    name = record.get("name")
    name = name if isinstance(name, str) and name.strip() else fallback_ref

    normalized: Dict[str, Any] = {
        **record,
        "kind": kind,
        "id": normalized_id,
        "ref": ref,
        "runtimeKind": runtime_kind,
        "name": name,
        "transform": transform,
    }
    if capabilities is not None:
        normalized["capabilities"] = capabilities
    else:
        normalized.pop("capabilities", None)
    return normalized


def _normalize_store(raw_store: Any, world_id: Any) -> Optional[Dict[str, Any]]:
    record = _as_record(raw_store)
    objects_record = _as_record(record.get("objects")) if record else None
    if not record or objects_record is None:
        return None

    objects: Dict[str, Any] = {}
    for ref, raw_object in objects_record.items():
        normalized = _normalize_game_object(raw_object, ref, world_id)
        if not normalized:
            continue
        normalized_ref = normalized["ref"] if isinstance(normalized.get("ref"), str) else ref
        objects[normalized_ref] = normalized

    schema_version = _to_number(record.get("schemaVersion"))
    store: Dict[str, Any] = {
        "schemaVersion": int(schema_version) if schema_version is not None else GAME_OBJECT_STORE_SCHEMA_VERSION,
        "objects": objects,
    }
    meta = _as_record(record.get("meta"))
    if meta is not None:
        store["meta"] = meta
    return store


def build_inventory_item_object(
    world_id: Any,
    item_id: str,
    quantity: Any,
    metadata: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Build a canonical item-kind GameObject (mirror of the TS builder)."""
    obj_id = item_id.strip()
    qty_num = _to_number(quantity)
    qty = max(0, int(qty_num)) if qty_num is not None else 0
    meta = _as_record(metadata) or {}
    name = meta.get("name")
    name = name if isinstance(name, str) and name.strip() else obj_id
    extra_item_data = _as_record(meta.get("itemData")) or {}
    reserved = {"name", "itemData", "id", "itemId", "qty", "quantity"}
    rest_meta = {k: v for k, v in meta.items() if k not in reserved}
    return {
        "kind": "item",
        "id": obj_id,
        "ref": f"item:{obj_id}",
        "name": name,
        "runtimeKind": "item",
        "transform": _create_fallback_transform(world_id),
        "capabilities": [{"id": "inventory_item", "enabled": True}],
        "itemData": {**rest_meta, **extra_item_data, "itemDefId": obj_id, "quantity": qty},
        "meta": {"source": "canonical.inventory"},
    }


# ---------------------------------------------------------------------------
# Public read API
# ---------------------------------------------------------------------------

def get_session_game_object_store(
    session_flags: Dict[str, Any], world_id: Any
) -> Dict[str, Any]:
    flags = _as_record(session_flags) or {}
    canonical = _normalize_store(flags.get("gameObjects"), world_id)
    return canonical or {"schemaVersion": GAME_OBJECT_STORE_SCHEMA_VERSION, "objects": {}}


def _matches_query(
    obj: Dict[str, Any],
    kind: Optional[str],
    capability: Optional[str],
    tags: Optional[List[str]],
    location_id: Optional[int],
) -> bool:
    if kind and obj.get("kind") != kind:
        return False
    if location_id is not None:
        obj_location = _to_number((obj.get("transform") or {}).get("locationId"))
        if obj_location is None or int(obj_location) != location_id:
            return False
    if capability and capability.strip():
        required = capability.strip()
        caps = obj.get("capabilities") or []
        if not any(c.get("id") == required and c.get("enabled") is not False for c in caps):
            return False
    if tags:
        owned = {t.lower() for t in _to_string_array(obj.get("tags"))}
        if not all(t.lower() in owned for t in tags):
            return False
    return True


def list_session_game_objects(
    session_flags: Dict[str, Any],
    world_id: Any,
    *,
    kind: Optional[str] = None,
    capability: Optional[str] = None,
    tags: Optional[List[str]] = None,
    location_id: Optional[int] = None,
) -> List[Dict[str, Any]]:
    store = get_session_game_object_store(session_flags, world_id)
    ordered = [store["objects"][ref] for ref in sorted(store["objects"].keys())]
    return [obj for obj in ordered if _matches_query(obj, kind, capability, tags, location_id)]


def get_session_game_object(
    session_flags: Dict[str, Any], world_id: Any, lookup: GameObjectLookup
) -> Optional[Dict[str, Any]]:
    store = get_session_game_object_store(session_flags, world_id)
    if isinstance(lookup, str):
        ref = lookup
    else:
        ref = to_game_object_ref(lookup.get("kind"), lookup.get("id"))
    return store["objects"].get(ref)


# ---------------------------------------------------------------------------
# Public write API (mutates session_flags in place, matching backend services)
# ---------------------------------------------------------------------------

def _write_store_objects(
    session_flags: Dict[str, Any],
    merged_objects: Dict[str, Any],
    base_store: Dict[str, Any],
) -> Dict[str, Any]:
    """Write the merged object set as the canonical store.

    As of the canonical cutover, the legacy ``flags.inventory.items`` mirror is
    no longer maintained — InventoryService and all readers consume canonical
    item GameObjects directly.
    """
    session_flags["gameObjects"] = {
        "schemaVersion": GAME_OBJECT_STORE_SCHEMA_VERSION,
        "objects": merged_objects,
        "meta": {
            **(_as_record(base_store.get("meta")) or {}),
            "updatedAt": datetime.now(timezone.utc).isoformat(),
        },
    }
    return session_flags


def upsert_session_game_objects(
    session_flags: Dict[str, Any], world_id: Any, objects: List[Dict[str, Any]]
) -> Dict[str, Any]:
    if not objects:
        return session_flags
    base = get_session_game_object_store(session_flags, world_id)
    merged = dict(base["objects"])
    for obj in objects:
        try:
            fallback_ref = to_game_object_ref(obj.get("kind"), obj.get("id"))
        except ValueError:
            continue
        normalized = _normalize_game_object(obj, fallback_ref, world_id)
        if not normalized:
            continue
        key = normalized["ref"] if isinstance(normalized.get("ref"), str) else fallback_ref
        merged[key] = normalized
    return _write_store_objects(session_flags, merged, base)


def remove_session_game_objects(
    session_flags: Dict[str, Any], world_id: Any, refs: List[str]
) -> Dict[str, Any]:
    if not refs:
        return session_flags
    base = get_session_game_object_store(session_flags, world_id)
    merged = dict(base["objects"])
    changed = False
    for ref in refs:
        if ref in merged:
            del merged[ref]
            changed = True
    if not changed:
        return session_flags
    return _write_store_objects(session_flags, merged, base)


# ---------------------------------------------------------------------------
# NPC component helpers
# ---------------------------------------------------------------------------
#
# Single-component get/set/remove on the canonical npc GameObject. Used by
# ecs_helpers (narrative component) and apply_stat_deltas (stats:<def_id>
# components) so both share one upsert-preserving-other-fields path.


def _npc_ref(npc_id: Any) -> str:
    if isinstance(npc_id, str):
        return npc_id if npc_id.startswith("npc:") else f"npc:{npc_id}"
    return f"npc:{npc_id}"


def get_npc_component(
    session_flags: Dict[str, Any],
    world_id: Any,
    npc_id: Any,
    component_type: str,
) -> Optional[Dict[str, Any]]:
    """Return the single component matching ``component_type`` on the canonical
    npc (with hydration), or ``None`` when absent."""
    obj = get_session_game_object(session_flags, world_id, _npc_ref(npc_id))
    if not obj:
        return None
    for comp in obj.get("components") or []:
        if isinstance(comp, dict) and comp.get("type") == component_type:
            return comp
    return None


def set_npc_component(
    session_flags: Dict[str, Any],
    world_id: Any,
    npc_id: Any,
    component_type: str,
    data: Dict[str, Any],
    enabled: bool = True,
) -> Dict[str, Any]:
    """Upsert a single component onto the canonical npc, preserving all other
    components and hydrated fields (transform, npcData, capabilities, tags, ...).

    Creates a minimal npc object if none exists yet."""
    obj = get_session_game_object(session_flags, world_id, _npc_ref(npc_id))
    current = list(obj.get("components") or []) if obj else []
    components = [c for c in current if c.get("type") != component_type]
    components.append({"type": component_type, "enabled": enabled, "data": dict(data)})
    if obj is None:
        normalized_id = npc_id
        if isinstance(npc_id, str) and npc_id.startswith("npc:"):
            try:
                normalized_id = int(npc_id[4:])
            except (TypeError, ValueError):
                normalized_id = npc_id[4:]
        payload: Dict[str, Any] = {
            "kind": "npc",
            "id": normalized_id,
            "name": f"NPC {normalized_id}",
            "components": components,
        }
    else:
        payload = {**obj, "components": components}
    return upsert_session_game_objects(session_flags, world_id, [payload])


def remove_npc_component(
    session_flags: Dict[str, Any],
    world_id: Any,
    npc_id: Any,
    component_type: str,
) -> Dict[str, Any]:
    """Remove a component from the canonical npc by type (no-op when absent)."""
    obj = get_session_game_object(session_flags, world_id, _npc_ref(npc_id))
    if not obj:
        return session_flags
    current = obj.get("components") or []
    if not any(c.get("type") == component_type for c in current):
        return session_flags
    payload = {**obj, "components": [c for c in current if c.get("type") != component_type]}
    return upsert_session_game_objects(session_flags, world_id, [payload])


# ---------------------------------------------------------------------------
# Convenience accessors (item / npc / component)
# ---------------------------------------------------------------------------

def item_quantity(obj: Dict[str, Any]) -> int:
    if obj.get("kind") != "item":
        return 0
    item_data = _as_record(obj.get("itemData")) or {}
    quantity = _to_number(item_data.get("quantity"))
    return max(0, int(quantity)) if quantity is not None else 0


def get_component(obj: Dict[str, Any], component_type: str) -> Optional[Dict[str, Any]]:
    required = component_type.strip() if isinstance(component_type, str) else ""
    if not required:
        return None
    for comp in obj.get("components") or []:
        if isinstance(comp, dict) and comp.get("type") == required:
            return comp
    return None


def has_capability(obj: Dict[str, Any], capability_id: str) -> bool:
    required = capability_id.strip() if isinstance(capability_id, str) else ""
    if not required:
        return False
    for cap in obj.get("capabilities") or []:
        if isinstance(cap, dict) and cap.get("id") == required and cap.get("enabled") is not False:
            return True
    return False
